ROVER-135 Make paths in `supergraph.yaml` resolve relative to the location of the file, not the current working directory of Rover

jonathanrainer opened the pull request:
As per the title, to support integration with the Language Server we need to remove a 'quirk' of Rover whereby if you specify a file in the `supergraph.yaml` with a relative path, that path is resolved relative to the location that `rover` is running in at the time. This doesn't work in that context so instead we are changing it to resolve relative to the location of the `supergraph.yaml` file itself.

I've added a unit test for this and run my own tests and it all seems to work ok. 

In addition I've fixed a test that has been broken for a long time, a more detailed write up on this test will follow.

Green Smoke Test Run: https://github.com/apollographql/rover/actions/runs/10735462842

jonathanrainer committed:
ROVER-135 Fix behaviour w.r.t erroring

jonathanrainer committed:
ROVER-135 Actually fix the test itself

jonathanrainer committed:
ROVER-135 Fix test that was broken for a long time

jonathanrainer commented on Cargo.toml:
This is not to be a permanent change, if we want to go down this route I'll get a PR merged into `federation-rs` before we merge this PR

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -89,9 +92,20 @@ pub async fn get_supergraph_config(
+            // to ensure we maintain correct semantics
+            config = match file_descriptor {
+                FileDescriptorType::Stdin => {
+                    let current_dir = Utf8PathBuf::try_from(current_dir()?)?;
+                    correctly_resolve_paths(config, &current_dir)?
```
This behaviour could be up for debate, to me it feels sensible that if someone pipes in a `supergraph.yaml` file then the directories should resolve relative to the current working directory of `rover` because I struggle to see an alternative. However I appreciate that might be a bit surprising, so an open to alternative suggestions!

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -108,264 +122,73 @@ pub async fn get_supergraph_config(
     Ok(supergraph_config)
 }
 
-/// Merge local and remote supergraphs, making sure that the federation version is correct: eg, when
```
This looks like a lot of change but I'm just moving code around so that the test code is near the implementing code

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -108,264 +122,73 @@ pub async fn get_supergraph_config(
-            .get_federation_version()
-            .unwrap_or_else(|| FederationVersion::LatestFedTwo)
-    })
+fn correctly_resolve_paths(
```
This is the meat of the change, essentially we post-process the supergraph config and canonicalise all the paths. The reason it worked as it did before was because it maintained the relative paths all the way through, until it resolved them very near the end of the process, so here we make it very explicit what path we actually want and error if we can't find it.

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -498,6 +318,15 @@ mod test_get_supergraph_config {
+                                    sdl
+                                  }
+                                }
+                              }
```
This is to fix up these tests, for which failures were being obscured because of the problems with `telemetry.rs`

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -638,40 +467,349 @@ mod test_get_supergraph_config {
+
+    #[rstest]
+    #[tokio::test]
+    async fn test_file_paths_become_canonicalised_on_read(
```
This is the new test which checks we canonicalise the paths on read so they are consistent throughout

jonathanrainer commented on src/utils/telemetry.rs:
```diff
@@ -151,15 +151,15 @@ mod tests {
 
     #[test]
     fn it_can_serialize_commands_with_arguments() {
-        let args = vec![PKG_NAME, "config", "list", "--help"];
```
This test was causing our test runner to not execute every test in CI, or when running locally, look out for a write up on this in the near future

jonathanrainer committed:
ROVER-135 Fix tests running on Windows

It appears these particular tests were relying on specific Windows
line breaks etc. This is very error-prone, and in reality unnecessary
as we can match on a smaller subset of the query itself and still
get sensible answers.

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -444,11 +267,6 @@ mod test_get_supergraph_config {
         Some(String::from("pandas")),
         Some(vec![(String::from("pandas"), String::from("local"))]),
     )]
-    #[case::local_takes_precedence(
-        Some(String::from("pandas")),
-        Some(String::from("pandas")),
-        Some(vec![(String::from("pandas"), String::from("local"))]),
-    )]
```
Why this was duplicated I'm not sure, but has now been removed

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -466,6 +284,12 @@ mod test_get_supergraph_config {
             let graphref_raw = format!("{name}@{variant}");
             let url = format!("http://{}.remote.com", name);
             server.mock(|when, then| {
+                let request_body_partial = json!({
+                    "variables": {
+                        "graph_ref": graphref_raw,
+                    },
+                    "operationName": "SubgraphFetchAllQuery"
+                });
```
This replaces the previous contents of L487 - 514. It seems that on Windows the line breaks that were contained within the query expression were causing the tests to trip up. There's no need for this because we can use another function of the Mock library to match on a partial JSON body, that still could only have been sent from the bit of the code we're testing.

A similar change is made in another test that fell foul of this issue.

jonathanrainer commented on src/utils/supergraph_config.rs:
```diff
@@ -484,71 +308,7 @@ mod test_get_supergraph_config {
-                    .json_body(body);
-            });
-
-            server.mock(|when, then| {
-                let body = json!({
-                  "data": {
-                    "graph": {
-                      "variant": {
-                        "subgraphs": [
-                          {
-                            "name": name
-                          }
-                        ]
-                      }
-                    }
-                  }
-                });
-                when.method(httpmock::Method::POST)
-                    .path("/")
-                    .json_body_obj(&json!({
-                        "query": indoc!{
-                          r#"
-                      query IsFederatedGraph($graph_id: ID!, $variant: String!) {
-                        graph(id: $graph_id) {
-                          variant(name: $variant) {
-                            subgraphs {
-                              name
-                            }
-                          }
-                        }
-                      }
-                      "#
-                        },
-                        "variables": {
-                            "graph_id": name,
-                            "variant": "current"
-                        },
-                        "operationName": "IsFederatedGraph"
-                    }));
```
This bit of code is now unnecessary because of Glasser's refactor a while back. Therefore it's just adding lines for no good reason

jonathanrainer reviewed (commented):
Have added a few explainer comments throughout, it might be helpful to go commit by commit but hopefully these help to explain

loshz commented on src/error/metadata/suggestion.rs:
```diff
@@ -1,10 +1,9 @@
-
+use crate::utils::env::RoverEnvKey;
+use camino::Utf8PathBuf;
 use rover_client::shared::GraphRef;
 use rover_std::Style;
-
-use crate::utils::env::RoverEnvKey;
+use std::cmp::Ordering;
+use std::fmt::{self, Display, Write as _};
 
 use serde::Serialize;
```
Imports look wonky here:
- `std` should be at the top.
- 3rd party second.
- `crate/super` last.

```rust
use std::cmp::Ordering;
use std::fmt::{self, Display, Write as _};

use camino::Utf8PathBuf;
use rover_client::shared::GraphRef;
use rover_std::Style;
use serde::Serialize;

use crate::utils::env::RoverEnvKey;
```

jonathanrainer committed:
ROVER-135 Working version of resolution behaviour

Still needs tests and will need a federation-rs PR merging
to get it all aligned.

jonathanrainer committed:
ROVER-135 Add new tests for paths

jonathanrainer committed:
ROVER-135 Resolving odd rebase
